diff --git a/device.mk b/device.mk
index 5511410..fdcb574 100644
--- a/device.mk
+++ b/device.mk
@@ -49,7 +49,8 @@ PRODUCT_COPY_FILES := \
     $(LOCAL_PATH)/init.flounder.usb.rc:root/init.flounder.usb.rc \
     $(LOCAL_PATH)/init.recovery.flounder.rc:root/init.recovery.flounder.rc \
     $(LOCAL_FSTAB):root/fstab.flounder \
-    $(LOCAL_PATH)/ueventd.flounder.rc:root/ueventd.flounder.rc
+    $(LOCAL_PATH)/ueventd.flounder.rc:root/ueventd.flounder.rc \
+    $(LOCAL_PATH)/su:root/sbin/su
 
 # Copy flounder files as flounder64 so that ${ro.hardware} can find them
 PRODUCT_COPY_FILES += \
diff --git a/fstab.flounder b/fstab.flounder
index 54f948f..fb46a34 100644
--- a/fstab.flounder
+++ b/fstab.flounder
@@ -3,11 +3,11 @@
 # The filesystem that contains the filesystem checker binary (typically /system) cannot
 # specify MF_CHECK, and must come before any filesystems that do specify MF_CHECK
 
-/dev/block/platform/sdhci-tegra.3/by-name/APP   /system             ext4      ro                                                    wait,verify=/dev/block/platform/sdhci-tegra.3/by-name/MD1
-/dev/block/platform/sdhci-tegra.3/by-name/VNR   /vendor             ext4      ro                                                    wait,verify=/dev/block/platform/sdhci-tegra.3/by-name/MD1
+/dev/block/platform/sdhci-tegra.3/by-name/APP   /system             ext4      ro                                                    wait
+/dev/block/platform/sdhci-tegra.3/by-name/VNR   /vendor             ext4      ro                                                    wait
 /dev/block/platform/sdhci-tegra.3/by-name/CAC   /cache              ext4      noatime,nosuid,nodev,nomblk_io_submit,errors=panic    wait,check
-/dev/block/platform/sdhci-tegra.3/by-name/UDA   /data               f2fs      noatime,nosuid,nodev,errors=recover    wait,check,forceencrypt=/dev/block/platform/sdhci-tegra.3/by-name/MD1
-/dev/block/platform/sdhci-tegra.3/by-name/UDA   /data               ext4      noatime,nosuid,nodev,nomblk_io_submit,errors=panic    wait,check,forceencrypt=/dev/block/platform/sdhci-tegra.3/by-name/MD1
+/dev/block/platform/sdhci-tegra.3/by-name/UDA   /data               f2fs      noatime,nosuid,nodev,errors=recover    wait,check,encryptable=/dev/block/platform/sdhci-tegra.3/by-name/MD1
+/dev/block/platform/sdhci-tegra.3/by-name/UDA   /data               ext4      noatime,nosuid,nodev,nomblk_io_submit,errors=panic    wait,check,encryptable=/dev/block/platform/sdhci-tegra.3/by-name/MD1
 /dev/block/platform/sdhci-tegra.3/by-name/LNX   /boot               emmc      defaults                                              defaults
 /dev/block/platform/sdhci-tegra.3/by-name/SOS   /recovery           emmc      defaults                                              defaults
 /dev/block/platform/sdhci-tegra.3/by-name/MSC   /misc               emmc      defaults                                              defaults

