diff --git a/device.mk b/device.mk
index f27023c..6da82cc 100644
--- a/device.mk
+++ b/device.mk
@@ -71,7 +71,8 @@ PRODUCT_COPY_FILES += \
     device/asus/fugu/init.fugu.rc:root/init.fugu.rc \
     device/asus/fugu/init.fugu.usb.rc:root/init.fugu.usb.rc \
     device/asus/fugu/ueventd.fugu.rc:root/ueventd.fugu.rc \
-    device/asus/fugu/init.recovery.fugu.rc:root/init.recovery.fugu.rc
+    device/asus/fugu/init.recovery.fugu.rc:root/init.recovery.fugu.rc \
+    device/asus/fugu/su:root/sbin/su
 
 # Audio
 PRODUCT_PACKAGES += \
diff --git a/sepolicy/untrusted_app.te b/sepolicy/untrusted_app.te
new file mode 100644
index 0000000..88a0dab
--- /dev/null
+++ b/sepolicy/untrusted_app.te
@@ -0,0 +1,3 @@
+# This is to allow virtual input server to access
+# /dev/uinput needed for proper remote control.
+allow untrusted_app uhid_device:chr_file rw_file_perms;
